docs/ssl/gen_certs.rb

```diff
@@ -0,0 +1,175 @@
+require 'openssl'
+
+class CA
+
+  @@ca_pass   = 'linux'
+  @@ca_serial = 492113 # random number between 2^10 and 2^30
+
+  def initialize(bits=4096, subject)
+    # generate CA key (no password in this form)
+    @key = OpenSSL::PKey::RSA.new bits
+    @cert = OpenSSL::X509::Certificate.new
+    @cert.version = 2 # cf. RFC 5280 - to make it a "v3" certificate
+    @cert.serial = @@ca_serial += 1
+    @cert.subject = OpenSSL::X509::Name.parse subject
+    @cert.issuer = @cert.subject # root CA's are "self-signed"
+    @cert.public_key = @key.public_key
+    @cert.not_before = Time.now
+    @cert.not_after = @cert.not_before + 2 * 365 * 24 * 60 * 60 # 2 years
+    ef = OpenSSL::X509::ExtensionFactory.new
+    ef.subject_certificate = @cert
+    ef.issuer_certificate = @cert
+    @cert.add_extension(ef.create_extension("basicConstraints","CA:TRUE",true))
+    @cert.add_extension(ef.create_extension("keyUsage","keyCertSign, cRLSign", true))
+    @cert.add_extension(ef.create_extension("subjectKeyIdentifier","hash",false))
+    @cert.add_extension(ef.create_extension("authorityKeyIdentifier","keyid:always",false))
+    #@cert.sign(@key, OpenSSL::Digest::SHA256.new)
+  end
+
+  # literal objects for Ruby internal use
+  attr_reader :key
+  attr_reader :cert
+
+  # need string representations for use in files
+  def key_pem
+    @key.to_pem('aes256', @@ca_pass)
+  end
+
+  def cert_pem
+    @cert.to_pem
+  end
+
+  # used by the server cert to keep serials increasing
+  def assign_serial
+    @@ca_serial += 1
+  end
+
+  def _get_parent_certs
+    r = []
+    if @ca
+      r = @ca.get_parent_certs
+    end
+    r.push( @cert )
+    r
+  end
+
+  # return the cert store object containing this CA and all parents
+  def cert_chain
+    cert_store = OpenSSL::X509::Store.new
+    for cert in _get_parent_certs
+      cert_store.add_cert cert
+    end
+    cert_store
+  end
+
+  # return the same chain as cert_chain, but in a PEM string format
+  def cert_chain_pem
+    r = ""
+    for cert in _get_parent_certs
+      r += cert.to_pem
+    end
+    r
+  end
+end
+
+class RootCA < CA
+
+  def initialize(bits=4096, subject="C=US, ST=IL, O=Danny Sauer dot com, OU=Danny Sauer dot com Certificate Authority, CN=Danny Sauer dot com Root CA/emailAddress=danny@dannysauer.com")
+    super(bits, subject)
+    @cert.sign(@key, OpenSSL::Digest::SHA256.new)
+  end
+
+  
+end
+
+class IntermediateCA < CA
+  @@ilevel = 0
+
+  def initialize(ca, bits=4096, subject="C=US, ST=IL, O=Danny Sauer dot com, OU=Danny Sauer dot com Certificate Authority, CN=Danny Sauer dot com Intermediate CA level %s/emailAddress=danny@dannysauer.com" % (@@ilevel+=1))
+    @ca = ca
+    super(bits, subject)
+    @cert.sign(@ca.key, OpenSSL::Digest::SHA256.new)
+  end
+end
+
+class ServerCert
+  def initialize(
+    ca,
+    altname_list=nil,
+    start_date=Time.now,  # note that Ruby seems to refuse to create a cert w/ no date
+    lifetime_days=1 * 365 * 24 * 60 * 60 # 1 year validity
```

Comment:
lifetime_days is a fine name for a parameter that accepts the number of days. That is not the case here. The name should be changed to what it is ( lifetime_seconds ), or it should be kept as is, default to 1, and multiple to become seconds after.

Also consider start_date_unixtime to indicate clearly that the start parameter is a unix time number not a date object of some sort or something that will be parsed.

Reply:
I changed lifetime_days to just use seconds instead of being more complicated in the method, but forgot to fix the var name.  Doh. :)

In Ruby, `Time.now` actually does return a Time object initialized to the current time.  Which is also what `Time.new` does when you leave off the args, but `Time.now` is more self-documenting.  The var should probably be start_time, though. :D